source/game/kart/KartState.cc

```diff
@@ -212,45 +213,64 @@ void KartState::calcCollisions() {
 
     m_wallBonkTimer = std::max(0, m_wallBonkTimer - 1);
 
-    if (colData.bWall) {
-        m_bWallCollision = true;
+    bool hwg = false;
+
+    if (colData.bWall || colData.bWall3) {
+        if (colData.bWall) {
+            m_bWallCollision = true;
+        }
+
+        if (colData.bWall3) {
+            m_bWall3Collision = true;
+        }
 
         if (!wasWallCollision) {
             m_bWallCollisionStart = true;
         }
 
         m_wallBonkTimer = 2;
-    }
 
-    if (m_hwgTimer == 0 && colData.movement.y > 1.0f) {
-        EGG::Vector3f movement = colData.movement;
-        movement.normalise();
-
-        if (movement.dot(EGG::Vector3f::ey) > 0.8f &&
-                colData.floorNrm.dot(EGG::Vector3f::ey) > 0.85f &&
-                movement.x * colData.wallNrm.x + movement.z * colData.wallNrm.z < 0.0f) {
-            colData.wallNrm.y = 0.0f;
-            colData.wallNrm.normalise();
-            wallNrm = colData.wallNrm;
+        if (m_hwgTimer == 0 && colData.movement.y > 1.0f) {
+            EGG::Vector3f movement = colData.movement;
+            movement.normalise();
+
+            if (movement.dot(EGG::Vector3f::ey) > 0.8f &&
+                    colData.wallNrm.dot(EGG::Vector3f::ey) > 0.85f &&
+                    (movement.x * colData.wallNrm.x + movement.z * colData.wallNrm.z < 0.0f ||
+                            collide()->colPerpendicularity() >= 1.0f)) {
+                colData.wallNrm.y = 0.0f;
+                colData.wallNrm.normalise();
+                wallNrm = colData.wallNrm;
+
+                if (wallNrm.length() < 0.05f) {
+                    wallNrm = movement;
+                    wallNrm.y = 0.0f;
+                }
 
-            if (wallNrm.length() < 0.05f) {
-                wallNrm = movement;
-                wallNrm.y = 0.0f;
+                hwg = true;
             }
         }
     }
 
-    if (softWallCount > 0) {
+    if (softWallCount > 0 || hwg) {
         m_bUNK2 = true;
         m_softWallSpeed = wallNrm;
         m_softWallSpeed.normalise();
         if (!m_bHop) {
             m_bSoftWallDrift = true;
         }
 
-        if (hitboxGroupSoftWallCollision || isBike()) {
+        if (hwg) {
+            m_bHWG = true;
+        }
+
+        if (hitboxGroupSoftWallCollision || hwg || isBike()) {
             m_bSomethingWallCollision = true;
             m_hwgTimer = 10;
+
+            if (hwg) {
+                m_hwgTimer <<= 1;
```

Comment:
```suggestion
                m_hwgTimer *= 2;
```
The bit-shifting is a compiler optimization for multiplying against a literal.